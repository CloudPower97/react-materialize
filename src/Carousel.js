import React, { Children, cloneElement, useEffect, useRef } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

const Carousel = ({
  children,
  className,
  fixedItem,
  images,
  centerImages,
  options,
  ...props
}) => {
  const _carousel = useRef(null);
  const elemsToRender = children || images || [];

  useEffect(() => {
    const instance = M.Carousel.init(_carousel.current, options);

    return () => {
      instance && instance.destroy();
    };
  }, [_carousel]);

  const renderItems = child => {
    if (typeof child === 'string') {
      return (
        <a
          className={cx('carousel-item', {
            'valign-wrapper': centerImages
          })}
        >
          <img src={child} alt="" />
        </a>
      );
    }

    return cloneElement(child, {
      className: cx('carousel-item', child.props.className, {
        'valign-wrapper': centerImages
      })
    });
  };

  return (
    elemsToRender && (
      <div
        ref={_carousel}
        className={cx(
          'carousel',
          { 'carousel-slider': options.fullWidth },
          className
        )}
        {...props}
      >
        {Children.map(elemsToRender, renderItems)}
        {fixedItem && (
          <div className="carousel-fixed-item center">{fixedItem}</div>
        )}
      </div>
    )
  );
};

Carousel.propTypes = {
  /*
   * Children to render as carousel elements
   */
  children: PropTypes.any,
  /*
   * Array of image url's
   */
  images: PropTypes.arrayOf(PropTypes.string),
  /*
   * Makes the images centered inside the carousel using '.valign-wrapper' CSS helper
   */
  centerImages: PropTypes.bool,
  /*
   * Fixed element on slider
   */
  fixedItem: PropTypes.node,
  /*
   * Additional classNames for carousel wrapper
   */
  className: PropTypes.string,
  /*
   * Carousel ID for jQuery manipulating
   */
  carouselId: PropTypes.string,
  /*
   * Carousel initialization options
   * <a href="http://materializecss.com/carousel.html">http://materializecss.com/carousel.html</a>
   */
  options: PropTypes.shape({
    /*
     * Transition duration in milliseconds. (Default: 200)
     */
    duration: PropTypes.number,
    /*
     * Perspective zoom. If 0, all items are the same size. (Default: -100)
     */
    dist: PropTypes.number,
    /*
     * Set the spacing of the center item. (Default: 0)
     */
    shift: PropTypes.number,
    /*
     * Set the padding between non center items. (Default: 0)
     */
    padding: PropTypes.number,
    /*
     * Set the number of visible items. (Default: 5)
     */
    numVisible: PropTypes.number,
    /*
     * Make the carousel a full width slider. (Default: false)
     */
    fullWidth: PropTypes.bool,
    /*
     * Set to true to show indicators. (Default: false)
     */
    indicators: PropTypes.bool,
    /*
     * Don't wrap around and cycle through items. (Default: false)
     */
    noWrap: PropTypes.bool,
    /*
     * Callback for when a new slide is cycled to. (Default: null)
     */
    onCycleTo: PropTypes.func
  })
};

Carousel.defaultProps = {
  options: {
    duration: 200,
    dist: -100,
    shift: 0,
    padding: 0,
    numVisible: 5,
    fullWidth: false,
    indicators: false,
    noWrap: false,
    onCycleTo: null
  }
};

export default Carousel;
