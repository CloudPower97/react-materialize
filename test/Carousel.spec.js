import React from 'react';
import { shallow, mount } from 'enzyme';
import Carousel from '../src/Carousel';
import mocker from './helper/new-mocker';

describe('<Carousel />', () => {
  let wrapper;

  const images = [
    'https://lorempixel.com/250/250/nature/1',
    'https://lorempixel.com/250/250/nature/2',
    'https://lorempixel.com/250/250/nature/3',
    'https://lorempixel.com/250/250/nature/4'
  ];

  test('renders', () => {
    wrapper = shallow(<Carousel images={images} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders fixed items', () => {
    const fixedItem = <span>Do you rock!?</span>;
    wrapper = shallow(<Carousel images={images} fixedItem={fixedItem} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders centered images', () => {
    wrapper = shallow(<Carousel images={images} centerImages />);
    expect(wrapper).toMatchSnapshot();
  });

  test('handles content slides', () => {
    const child = (
      <div className="red">
        <h2>First Panel</h2>
        <p>This is your first panel</p>
      </div>
    );
    wrapper = shallow(<Carousel images={images}>{child}</Carousel>);
    expect(wrapper).toMatchSnapshot();
  });

  test('accepts className props', () => {
    wrapper = shallow(<Carousel images={images} className="center" />);
    expect(wrapper.hasClass('carousel center')).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  test('accepts external `id` value', () => {
    wrapper = shallow(<Carousel images={images} id="ID" />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('initialises', () => {
    const carouselInitMock = jest.fn();
    const carouselInstanceDestroyMock = jest.fn();
    const carouselMock = {
      init: (el, options) => {
        carouselInitMock(options);
        return {
          destroy: carouselInstanceDestroyMock
        };
      }
    };

    const restore = mocker('Carousel', carouselMock);

    beforeEach(() => {
      carouselInitMock.mockClear();
      carouselInstanceDestroyMock.mockClear();
    });

    afterAll(() => {
      restore();
    });

    test('uses default options if none are given', () => {
      wrapper = mount(<Carousel />);

      expect(carouselInitMock).toHaveBeenCalledWith({
        duration: 200,
        dist: -100,
        shift: 0,
        padding: 0,
        numVisible: 5,
        fullWidth: false,
        indicators: false,
        noWrap: false,
        onCycleTo: null
      });
    });

    test('handles full width sliders', () => {
      wrapper = mount(
        <Carousel images={images} options={{ fullWidth: true }} />
      );
      expect(wrapper).toMatchSnapshot();
      expect(carouselInitMock).toHaveBeenCalledWith({ fullWidth: true });
    });

    test('more options', () => {
      const options = { padding: 12, fullWidth: true, indicators: false };
      wrapper = mount(<Carousel images={images} options={options} />);
      expect(carouselInitMock).toHaveBeenCalledWith(options);
    });
  });
});
