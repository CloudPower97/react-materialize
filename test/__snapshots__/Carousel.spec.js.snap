// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Carousel /> accepts className props 1`] = `
<div
  className="carousel center"
>
  <a
    className="carousel-item"
    key=".0"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/1"
    />
  </a>
  <a
    className="carousel-item"
    key=".1"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/2"
    />
  </a>
  <a
    className="carousel-item"
    key=".2"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/3"
    />
  </a>
  <a
    className="carousel-item"
    key=".3"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/4"
    />
  </a>
</div>
`;

exports[`<Carousel /> accepts external \`id\` value 1`] = `
<div
  className="carousel"
  id="ID"
>
  <a
    className="carousel-item"
    key=".0"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/1"
    />
  </a>
  <a
    className="carousel-item"
    key=".1"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/2"
    />
  </a>
  <a
    className="carousel-item"
    key=".2"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/3"
    />
  </a>
  <a
    className="carousel-item"
    key=".3"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/4"
    />
  </a>
</div>
`;

exports[`<Carousel /> handles content slides 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-item red"
    key=".0"
  >
    <h2>
      First Panel
    </h2>
    <p>
      This is your first panel
    </p>
  </div>
</div>
`;

exports[`<Carousel /> initialises handles full width sliders 1`] = `
<Carousel
  images={
    Array [
      "https://lorempixel.com/250/250/nature/1",
      "https://lorempixel.com/250/250/nature/2",
      "https://lorempixel.com/250/250/nature/3",
      "https://lorempixel.com/250/250/nature/4",
    ]
  }
  options={
    Object {
      "fullWidth": true,
    }
  }
>
  <div
    className="carousel carousel-slider"
  >
    <a
      className="carousel-item"
      key=".0"
    >
      <img
        alt=""
        src="https://lorempixel.com/250/250/nature/1"
      />
    </a>
    <a
      className="carousel-item"
      key=".1"
    >
      <img
        alt=""
        src="https://lorempixel.com/250/250/nature/2"
      />
    </a>
    <a
      className="carousel-item"
      key=".2"
    >
      <img
        alt=""
        src="https://lorempixel.com/250/250/nature/3"
      />
    </a>
    <a
      className="carousel-item"
      key=".3"
    >
      <img
        alt=""
        src="https://lorempixel.com/250/250/nature/4"
      />
    </a>
  </div>
</Carousel>
`;

exports[`<Carousel /> renders 1`] = `
<div
  className="carousel"
>
  <a
    className="carousel-item"
    key=".0"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/1"
    />
  </a>
  <a
    className="carousel-item"
    key=".1"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/2"
    />
  </a>
  <a
    className="carousel-item"
    key=".2"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/3"
    />
  </a>
  <a
    className="carousel-item"
    key=".3"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/4"
    />
  </a>
</div>
`;

exports[`<Carousel /> renders centered images 1`] = `
<div
  className="carousel"
>
  <a
    className="carousel-item valign-wrapper"
    key=".0"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/1"
    />
  </a>
  <a
    className="carousel-item valign-wrapper"
    key=".1"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/2"
    />
  </a>
  <a
    className="carousel-item valign-wrapper"
    key=".2"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/3"
    />
  </a>
  <a
    className="carousel-item valign-wrapper"
    key=".3"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/4"
    />
  </a>
</div>
`;

exports[`<Carousel /> renders fixed items 1`] = `
<div
  className="carousel"
>
  <a
    className="carousel-item"
    key=".0"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/1"
    />
  </a>
  <a
    className="carousel-item"
    key=".1"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/2"
    />
  </a>
  <a
    className="carousel-item"
    key=".2"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/3"
    />
  </a>
  <a
    className="carousel-item"
    key=".3"
  >
    <img
      alt=""
      src="https://lorempixel.com/250/250/nature/4"
    />
  </a>
  <div
    className="carousel-fixed-item center"
  >
    <span>
      Do you rock!?
    </span>
  </div>
</div>
`;
